### PR TITLE
[Feature][UDF][WASM] support wasm udf

### LINKED ISSUE #20267: [Feature] Support Wasm UDF
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

## Backgroup

Doris currently supports both [Native UDF](https://doris.apache.org/zh-CN/docs/dev/ecosystem/udf/native-user-defined-function), [Remote UDF](https://doris.apache.org/zh-CN/docs/dev/ecosystem/udf/remote-user-defined-function) and [Java UDF](https://doris.apache.org/zh-CN/docs/dev/ecosystem/udf/java-user-defined-function) for user-defined functions.

Native UDF is written in C++ and has the best performance, but it is more difficult to write and debug, and may be limited by some system library versions(such as libc) that may not be compatible after upgrade.

Remote UDF solves the language problem very well, and in theory, UDF logic can be written in any language. But the disadvantage is that users need to implement their own high-performance UDF Service, and the efficiency is not good because of the RPC problem.

Java UDF is the main user-defined function solution in Doris. which reduces the migration cost of big data ecological users(Some big data ecologies such as Hive, Spark, etc. already exist a large number of ready-made  UDF). The implementation of Java UDF is to start the JVM and call the relevant UDF logic from the BE side through JNI.

## Motivation

Support for Wasm UDF is motivated by the following points:

1. Embeddable in [multiple programming languages. Users can write functional logic in multiple programming languages ​​(Rust, C/C++, Golang, Java, TypeScript, Haskell)
2. Secure by default. No file, network, or environment access, unless explicitly enabled.
3. High-performance. WebAssembly engine compiles bytecode into machine-native machine code for execution instead of interpreting it, which greatly improves execution efficiency and can achieve an efficiency close to native execution.

## How to implement

Need to start the Wasm Runtime and call the relevant UDF logic from the BE side by way of wasmtime-c-api.

## Scheduling
I have a preliminary plan to support Wasm UDF in Doris.

Phase I, I'll complete Create-Function-Statement for Wasm UDF.

Phase II, I'll add wasmtime lib and wasmtime-c-api to BE, and implement basic data type based on Wasm Basic ABI Type.

### Use case

_No response_

### Related issues

Remote UDF: https://github.com/apache/doris/pull/7519
Lua UDF: https://github.com/apache/doris/pull/5979
Java UDF: https://github.com/apache/doris/issues/8389

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by taptao**
## Proposed changes

Issue Number: close https://github.com/apache/doris/issues/20267

Support wasm udf via wasmtime library.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by dataroaring** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java [hunk 1]
```diff
@@ -490,6 +495,24 @@ private void analyzeJavaUdaf(String clazz) throws AnalysisException {
         }
     }
 
+    private void analyzeWasmUdf() throws AnalysisException {
```
Comment (zhangstar333) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java:498]: could add some example doc about create a wasm udf function,

Comment (taptao) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java:501]: done!

File: be/src/vec/functions/function_wasm.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+#include "util/wasm_manager.h"
+#include "vec/core/block.h"
+#include "vec/core/column_numbers.h"
+#include "vec/core/column_with_type_and_name.h"
+#include "vec/core/columns_with_type_and_name.h"
+#include "vec/core/types.h"
+#include "vec/data_types/data_type.h"
+#include "vec/functions/function.h"
+
+namespace doris {} // namespace doris
```
Comment (HappenLee) [inline be/src/vec/functions/function_wasm.h:38]: seems useless ? remove the line ?

File: be/src/vec/functions/function_wasm.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    Status open(FunctionContext* context, FunctionContext::FunctionStateScope scope) override;
+
+    Status execute(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                   size_t result, size_t input_rows_count, bool dry_run = false) override;
+
+    bool is_deterministic() const override { return false; }
+
+    bool is_deterministic_in_scope_of_query() const override { return false; }
+
+    bool is_use_default_implementation_for_constants() const override { return true; }
```
Comment (HappenLee) [inline be/src/vec/functions/function_wasm.h:79]: the default value is `true` no need override

File: be/src/vec/functions/function_wasm.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+Status FunctionWasm::open(FunctionContext* context, FunctionContext::FunctionStateScope scope) {
+    if (scope == FunctionContext::FRAGMENT_LOCAL) {
+        string local_location;
+        auto* function_cache = UserFunctionCache::instance();
+        RETURN_IF_ERROR(function_cache->get_watpath(_tfn.id, _tfn.hdfs_location, _tfn.checksum,
+                                                    &local_location));
+        //        ExecEnv::GetInstance()->wasm_function_manager();
+        std::shared_ptr<WasmFunctionManager> manager = std::make_shared<WasmFunctionManager>();
+        context->set_function_state(FunctionContext::THREAD_LOCAL, manager);
+        std::ifstream watFile;
```
Comment (HappenLee) [inline be/src/vec/functions/function_wasm.cpp:61]: better be code format like doris origin code:
`wat_file`
`str_string`

here i have a question? each time we need open to do `IO` task ?

File: be/src/vec/functions/function_wasm.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    auto result_nullable = return_type->is_nullable();
+    ColumnUInt8::MutablePtr null_map = nullptr;
+
+    if (result_nullable) {
+        return_type = remove_nullable(_return_type);
+        null_map = ColumnUInt8::create(input_rows_count, 0);
+        memset(null_map->get_data().data(), 0, input_rows_count);
+    }
+
+    //    ColumnPtr nested_column = nullptr;
```
Comment (HappenLee) [inline be/src/vec/functions/function_wasm.cpp:93]: if useless or dead code, please remove

File: be/src/vec/functions/function_wasm.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+        auto data_col = column.column->convert_to_full_column_if_const();
+
+        data_cols[arg_idx] = data_col;
+    }
+
+    // step1. process column value to wasm param
+    // step2. call wasm function
+    // step3. return wasm result to column value
+    int row_size = data_cols[0]->size();
```
Comment (HappenLee) [inline be/src/vec/functions/function_wasm.cpp:132]: ADD TODO: vec the code to call wasm fun

File: be/src/util/wasm_manager.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    wasmtime::Func func;
+    wasmtime::Instance instance;
+    WasmtimeRunInstance(const wasmtime::Func& func, const wasmtime::Instance& instance)
+            : func(func), instance(instance) {}
+};
+
+class WasmFunctionManager {
+private:
+    // wasmtime
+    wasmtime::Engine* engine;
```
Comment (HappenLee) [inline be/src/util/wasm_manager.h:42]: why not use `std::unique_ptr` ?

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java [hunk 8]
```diff
@@ -490,6 +498,24 @@ private void analyzeJavaUdaf(String clazz) throws AnalysisException {
...
+        try {
+            URL url = new URL(userFile);
+            try (Store<Void> store = Store.withoutData()) {
+                try (Engine engine = store.engine();
+                        Module module = Module.fromFile(engine, url.getFile())) {
+                    return;
```
[the remark this answers is not in the document]
Comment (taptao) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java:507]: I tried it before, but wasmtime-java jar doesn't support this method. I asked the wasmtime-java author this question, but there has been no reply so far.

File: be/src/vec/functions/function_wasm.h (hunk 3 above)
Comment (taptao) [reply be/src/vec/functions/function_wasm.h:79]: FunctionWasm extends IFunctionBase. IFunctionBase does not implement is_use_default_implementation_for_constants func. Did I misunderstand it?

File: be/src/vec/functions/function_wasm.cpp (hunk 4 above)
Comment (taptao) [reply be/src/vec/functions/function_wasm.cpp:61]: io task will be executed each time for now. but I only want to initialize it once, I don't know how to do it

File: be/src/util/wasm_manager.h (hunk 7 above)
Comment (taptao) [reply be/src/util/wasm_manager.h]: done

File: be/src/vec/functions/function_wasm.cpp (hunk 5 above)
Comment (taptao) [reply be/src/vec/functions/function_wasm.cpp]: done

File: be/src/vec/functions/function_wasm.h (hunk 2 above)
Comment (taptao) [reply be/src/vec/functions/function_wasm.h]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (zhangstar333) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/CreateFunctionStmt.java:507]: ok

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

**Comment by amorynan** [comment]
run buildall

**Comment by taptao** [comment]
reopen

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
Move to https://github.com/apache/doris/pull/30789 to view WASM PR

**Pull request closed without merging**